solution/q1/README.md

```diff
@@ -0,0 +1,224 @@
+index 335d8c2..cc1a850 100644
+--- a/solution/q1/terraform/vars.tf
++++ b/solution/q1/terraform/vars.tf
+@@ -2,3 +2,8 @@ variable "bucket_prefix" {
+   type    = string
+   default = "cd-test"
+ }
++
++variable "annalise_bucket_migrated" {
++  type    = bool
++  default = false
++}
+```
+
+</details>
+
+<details>
+
+<summary>Run `terraform import` of all resources related to that bucket (bucket itself, acl, public access, SSE...):</summary>
+
+```
+terraform import aws_s3_bucket.src_bucket annalise-ai-datalake
+terraform import aws_s3_bucket_acl.src_bucket_acl annalise-ai-datalake
+terraform import aws_s3_bucket_ownership_controls.src_bucket_oc annalise-ai-datalake
+terraform import aws_s3_bucket_server_side_encryption_configuration.src_bucket_sse annalise-ai-datalake
+terraform import aws_s3_bucket_public_access_block.src_bucket_access annalise-ai-datalake
+```
+
+</details>
+
+<details>
+
+<summary>Change `annalise_bucket_migrated` to `true`:</summary>
+
+```
+diff --git a/solution/q1/terraform/vars.tf b/solution/q1/terraform/vars.tf
+index 335d8c2..cc1a850 100644
+--- a/solution/q1/terraform/vars.tf
++++ b/solution/q1/terraform/vars.tf
+@@ -2,3 +2,8 @@ variable "bucket_prefix" {
+
+variable "annalise_bucket_migrated" {
+  type    = bool
+- default = false
++ default = true
+}
+```
+
+</details>
+
+## Utility script
+
+```
+cd util
+```
+
+To copy files between `annalise-ai-datalake` and `harrison-ai-landing` buckets, we use simple program
+in Go. It uses basic S3 API calls to list files in source bucket (`annalise-ai-datalake`) and copy them
+to destination (`harrison-ai-landing`). To accelerate the process, simple concurrency with channel
+and goroutines is used - `listS3Keys` produces the list of all keys into `keyChannel` which is later consumed
+by one or several goroutines (~threads) to do concurrent `copyS3Object` calls. There should be a reasonable
+limit to concurrency, local tests on 30G of dataset show that 4-core CPU can efficiently run 16 goroutines,
+without too much of CPU saturation (never reaches 90%) and more than twice faster throughput than default
+aws cli tools (`aws s3 sync`).
+
+To use the script, build it with docker and execute with AWS credentials passed as environment variables:
+```
+docker build -t util-q1:0.0.3 .
+docker run --rm -ti \
+  -e SRC_BUCKET=cd-test-annalise-ai-datalake \
+  -e DST_BUCKET=cd-test-harrison-ai-landing \
+  -e AWS_ACCESS_KEY_ID=$(grep aws_access_key_id ~/.aws/credentials | awk '{print $3}') \
+  -e AWS_SECRET_ACCESS_KEY=$(grep aws_secret_access_key ~/.aws/credentials | awk '{print $3}') \
+  util-q1:0.0.3
+```
+
+The script has verbose logging to standard output to track the progress - it can be useful when executed in
+managed environment, like Jenkins / RunDeck / GH actions / Octopus Deploy. If there was unexpected failure during execution, `START_KEY` can be used to continue from latest copied file (obtained from logging output
+of previous interrupted execution):
```

Comment:
@cdgz being able to recovery from interruption is definitely important here. I'm curious how this would interact with the concurrency in the go script, where IIUC you've got 16 goroutines concurrently processing items, and so the objects are not guaranteed to be processed in sequential order. Is there the potential for this recovery strategy to leave "holes" after an interruption is the objects are processed slightly out-of-order?

Reply:
Good point, my local tests were indeed with multiple threads and I did occasionally see this slight inconsistency when aborting the execution. To have a guaranteed consistency with "no holes" the better way is to rely on full and enumerated list of keys in the source bucket, which could be obtained via AWS Athena. Just as I mentioned in "Big amount of data" section of Question 2, we'd shard the full list and reorganize the goroutines to work in batches: say we take 1600 entries from the list and process them in parallel. Once batch is confirmed successful, batch ID is logged to the output and process moves on to next one. If there was interruption in between - calculate offset in the full list (based on batch ID) and restart from that batch.